Problem: no well-tested implementation of subscriptions module (fix #295)

yihuang opened the pull request:
Solution:
    - implement basic calendar and crontab logic
    - implement the module

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻


# PR Checklist:

- [x] Have you read the [CONTRIBUTING.md](https://github.com/crypto-com/chain-main/blob/master/CONTRIBUTING.md)?
- [x] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [x] Have you rebased your work on top of the latest master? 
- [x] Have you checked your code compiles? (`make`)
- [x] Have you included tests for any non-trivial functionality?
- [x] Have you checked your code passes the unit tests? (`make test`)
- [x] Have you checked your code formatting is correct? (`go fmt`)
- [x] Have you checked your basic code style is fine? (`golangci-lint run`)
- [x] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [x] If your changes affect the client infrastructure, have you run the integration test?
- [x] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-com/chain-main/blob/master/CHANGELOG.md)?
- [x] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-com/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-com/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/crypto-org-chain/chain-main/381)
<!-- Reviewable:end -->

yihuang commented:
TODO: add integration test

tomtau commented on third_party/proto/cosmos/base/abci/v1beta1/abci.proto:
```diff
@@ -0,0 +1,137 @@
+syntax = "proto3";
```
for the third party protos -- would it be possible with a submodule or a build script that will be pinned against the specific version-tag/commit sha on the external repositories?

yihuang replied:
We have to create submodule for the whole cosmos-sdk, can't have only the subdirectory.

tomtau replied:
that's ok -- the integration gRPC tests also cloned the whole repos

yihuang replied:
I removed the unneeded third party protos now, we actually only need the `pagination.proto`

tomtau commented on x/subscription/spec/03_messages.md:
````diff
@@ -46,11 +49,24 @@ ConsumeGas( count_period(create_time, expiration_time, plan.cron_spec, plan.cron
+The plan owner stop user's subscription.
+
+```protobuf
+message MsgCreateSubscription {
````
```suggestion
message MsgStopUserSubscription {
```
the message content should just have plan_owner + plan_id?

yihuang replied:
it should be `plan_owner` + `subscription_id`, updated spec.

devashishdxt commented on proto/subscription/v1beta1/subscription.proto:
```diff
@@ -0,0 +1,38 @@
+
+option go_package = "github.com/crypto-com/chain-main/x/subscription/types";
+
+message Params {
```
Some documentation on what these params mean. Or link to specs?

yihuang replied:
done

devashishdxt commented on x/subscription/client/cli/tx.go:
```diff
@@ -0,0 +1,258 @@
+	cmd.Flags().String(FlagTitle, "", "The plan title")
+	cmd.Flags().String(FlagDescription, "", "The plan description")
+	cmd.Flags().String(FlagPrice, "", "The plan price")
+	cmd.Flags().Uint32(flagDuration, 365*24*3600, "The plan subscription duration")
```
Unit of duration?

yihuang replied:
it's in seconds, I renamed it to make it clear.

tomtau commented:
there are a few conflicts with the latest master

yihuang commented:
> there are a few conflicts with the latest master

done

tomtau commented on x/subscription/types/cronspec.go:
```diff
@@ -0,0 +1,396 @@
+	case strings.Contains(s, "-"):
+		parts := strings.Split(s, "-")
+		if len(parts) == 2 || len(parts) == 3 {
+			start, err := strconv.Atoi(parts[0])
```
lgtm complains about these platform-specific functions -- could this use more precise / fixed with functions (`ParseUint`?)?

yihuang replied:
done

tomtau commented on go.mod:
```diff
@@ -17,6 +17,7 @@ require (
 	github.com/stretchr/testify v1.7.0
 	github.com/tendermint/tendermint v0.34.7
 	github.com/tendermint/tm-db v0.6.4
+	github.com/willf/bitset v1.1.11
```
ok, seems it's doing some unions and intersections, so perhaps beyond a simple bit array?

yihuang replied:
yes, maybe it's good to vendor it. it's basically an [one module package](https://github.com/willf/bitset/blob/master/bitset.go), I guess we don't need the optimized asm, just the generic version.

yihuang replied:
done, removed this dependency.

tomtau reviewed (changes requested):
overall ok, but I had to skip some parts (keeper, cronspec, ...), because I'm unsure what their correct behaviour should be.
Also this: https://github.com/crypto-org-chain/chain-main/issues/278 will be a blocker for this PR, because the code change is huge and there are some complex code parts and we should know what is exactly covered.

in terms of these being "well-tested", it should both test all features / message handling extensively as well as do [negative testing](https://en.wikipedia.org/wiki/Negative_testing), as there's a lot of signed vs unsigned, 32-bit vs 64-bit integer stuff etc. (it may be ok, but i skipped some parts and codecov warnings may not be precise at the moment)

JayT106 commented on proto/subscription/v1beta1/query.proto:
```diff
@@ -0,0 +1,87 @@
+syntax = "proto3";
+package subscription.v1beta1;
```
I'm not sure why the Cosmos SDK name it v1beta1. 
I will prefer to just call it v1 regardless of the package name versioning of the Cosmos SDK.

yihuang replied:
> 3. each module versions independently -- so subscriptions can be "v1" regardless of what mainnet base repo is and will only be bumped to "v2" if there are breaking changes in subscriptions?

+1

yihuang commented on app/genesis.go:
```diff
@@ -14,7 +16,6 @@ import (
 // NewDefaultGenesisState generates the default state for the application.
-func NewDefaultGenesisState() GenesisState {
-	encCfg := MakeEncodingConfig()
-	return ModuleBasics.DefaultGenesis(encCfg.Marshaler)
+func NewDefaultGenesisState(cdc codec.JSONMarshaler) GenesisState {
+	return ModuleBasics.DefaultGenesis(cdc)
```
it's to make it accept cdc as parameter, so the cdc can be shared in other places, it's used in `test_helpers.go` to setup test environment.

yihuang commented on x/subscription/types/msgs.go:
```diff
@@ -0,0 +1,188 @@
+
+// Constants pertaining to a contents
+const (
+	MaxDescriptionLength int = 5000
+	MaxTitleLength       int = 140
```
maybe, constant should be fine too I guess?

yihuang commented on app/app.go:
```diff
@@ -282,6 +291,7 @@ func New(
 	)
 
 	app.SupplyKeeper = supplykeeper.NewKeeper(appCodec, keys[supplytypes.StoreKey], app.BankKeeper, app.AccountKeeper)
+	app.SubscriptionKeeper = subscriptionkeeper.NewKeeper(appCodec, keys[subscriptiontypes.StoreKey], app.GetSubspace(subscriptiontypes.ModuleName), app.BankKeeper)
```
yeah, the codecov result always looks problematic to me.

yihuang replied:
```
$ go test -v -coverprofile=size_coverage.out ./... && go tool cover -html=size_coverage.out
```
Even run unit tests in the root directory like this, the report don't show the `New` function as covered, but it's definitely called in the unit tests.

tomtau replied:
maybe `covermode` needs to be set to atomic?

yihuang commented on x/subscription/types/msgs.go:
```diff
@@ -0,0 +1,188 @@
+		return sdkerrors.Wrap(ErrInvalidCronSpec, m.CronSpec.String())
+	}
+	if len(strings.TrimSpace(m.Title)) == 0 {
+		return sdkerrors.Wrap(ErrInvalidPlanContent, "proposal title cannot be blank")
```
grep and fixed similar issues.

yihuang commented on integration_tests/test_subscription.py:
```diff
@@ -0,0 +1,38 @@
+    subscription = cli.query_subscription(subscription_id)
+    print("subscription updated", subscription)
+    assert int(subscription["next_collection_time"]) > collection_time
+    assert cli.balance(subscriber) == amount - 10000000
```
done

yihuang commented on proto/subscription/v1beta1/tx.proto:
```diff
@@ -0,0 +1,60 @@
+    ];
+    uint32 duration_secs = 5;
+    CronSpec cron_spec = 6 [(gogoproto.nullable) = false];
+    uint32 tzoffset = 7;  // in seconds
```
fixed

yihuang commented on x/subscription/client/cli/query.go:
```diff
@@ -0,0 +1,231 @@
+package cli
```
https://github.com/crypto-org-chain/chain-main/pull/381/files#diff-1f05e4ecaa1ac6dc4321812f1e6c14b689a7e51505e16538799ff0da01f90700R786
it is testing the cli, which calls the grpc. But it don't test everyone of them.

yihuang commented on x/subscription/keeper/grpc_query.go:
```diff
@@ -0,0 +1,144 @@
+		}
+
+		matchPlanID := true
+		if req.PlanId >= 0 && subscription.PlanId != uint64(req.PlanId) {
```
because the query message use `int64` to express something like `Option<uint64>`, with negative values means not filtering.

yihuang commented on x/subscription/keeper/grpc_query.go:
```diff
@@ -0,0 +1,144 @@
+			return false, status.Error(codes.Internal, err.Error())
+		}
+
+		matchOwner := true
```
yes, `true` means not filtering by this condition.

yihuang commented on x/subscription/keeper/grpc_query.go:
```diff
@@ -0,0 +1,144 @@
+	return &types.QuerySubscriptionResponse{Subscription: subscription}, nil
+}
+
+// Votes returns single plan's votes
```
fixed all similar cases.

yihuang commented on x/subscription/keeper/keeper.go:
```diff
@@ -0,0 +1,411 @@
+
+func (k Keeper) PlanSubscriptonIterator(ctx sdk.Context, planID uint64) sdk.Iterator {
+	store := ctx.KVStore(k.storeKey)
+	return store.Iterator(types.PlanKeyPrefix, sdk.PrefixEndBytes(types.PlanKey(math.MaxUint64)))
```
Good catch, it's actually unused, the real one is the `SubscriptonPlanIterator`, remove it now.

akamsmrhill reviewed (commented):
n/a

yihuang commented on x/subscription/types/key.go:
```diff
@@ -0,0 +1,117 @@
+)
+
+// Keys for subscription store
+var (
```
added some comments into this, also a section in state spec.

yihuang commented on x/subscription/types/genesis.go:
```diff
@@ -0,0 +1,16 @@
+// Validate performs basic genesis state validation returning an error upon any
+// failure.
+func (gs GenesisState) Validate() error {
+	return nil
```
good catch, added several validations now.

yihuang commented on x/subscription/keeper/keeper.go:
```diff
@@ -0,0 +1,411 @@
+package keeper
```
added a section in state spec to explain the different types of keys in storage.

yihuang commented on x/subscription/types/cronspec.go:
```diff
@@ -0,0 +1,396 @@
+	set *bitset.BitSet
+	Min uint32
+	Max uint32
+}
```
this is explained in the `x/subscription/README.md` now.

tomtau commented on x/subscription/types/bitset.go:
```diff
@@ -0,0 +1,88 @@
+package types
```
will need to preserve the license/copyright notices: https://github.com/willf/bitset/blob/master/LICENSE
(both in the headers of the affected files + one entry in the NOTICE file)

yihuang replied:
done

yihuang commented on x/subscription/keeper/grpc_query.go:
```diff
@@ -0,0 +1,144 @@
+	ctx := sdk.UnwrapSDKContext(c)
+
+	plan, found := k.GetPlan(ctx, req.PlanId)
+	if !found {
```
it returns the plan as value, so it won't be nil.

yihuang commented on x/subscription/types/calendar.go:
```diff
@@ -0,0 +1,161 @@
+	return days*86400 + int64(tm.Hour)*3600 + int64(tm.Minute)*60 + int64(tm.Second)
+}
+
+// SecsToTM converts timestamp to TimeStruct
```
now it's all covered by unit tests.

yihuang commented on x/subscription/types/calendar.go:
```diff
@@ -0,0 +1,161 @@
+
+func Weekday(year, month, day int) int {
+	v := (DateToOrdinal(year, month, day) + 1) % 7
+	if v < 0 {
```
removed the negative branch, it's not needed, added input check at begining.

yihuang commented on x/subscription/types/codec.go:
```diff
@@ -0,0 +1,58 @@
+// RegisterLegacyAminoCodec registers all the necessary types and interfaces for the
+// subscription module.
+func RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
+	cdc.RegisterConcrete(&MsgCreatePlan{}, "chain-main/MsgCreatePlan", nil)
```
I'm not sure, it uses same conventions as cosmos-sdk here, [example](https://github.com/cosmos/cosmos-sdk/blob/master/x/ibc/applications/transfer/types/codec.go#L13).

JayT106 replied:
Is the link broken?

yihuang commented on proto/subscription/v1beta1/cronspec.proto:
```diff
@@ -0,0 +1,28 @@
+
+message CronItem {
+    oneof spec {
+        uint32 value = 1;
```
`Any` is the nil value, it's explained in `x/subscription/README.md`.

yihuang commented:
Local test coverage run shows:
```
keeper.go 87.8%
bitset.go 100%
calendar.go 95.2%
cronspec.go 91.9%
```
Logic code is all covered, only some validation error branches not.

yihuang commented on proto/subscription/v1beta1/subscription.proto:
```diff
@@ -0,0 +1,43 @@
+        (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins",
+        (gogoproto.moretags)     = "yaml:\"price\""
+    ];
+    uint32 duration_secs = 5;
```
updated the sign-ness in spec.

JayT106 commented on x/subscription/keeper/keeper.go:
```diff
@@ -0,0 +1,405 @@
+	}
+
+	k.SetPlan(ctx, plan)
+	k.SetPlanID(ctx, planID+1)
```
Why does the keeper always update planID to the latest planID + 1? What if we just store the ID of the latest plan in `SetPlan` and increase the ID counts only when creating the new plan
is any other function access `SetPlanID`?

Another idea in this function, Is it okay if we just pass the KVStore address instead of Context in the keeper's Get/Set function?   
In `CreatePlan` we can only get KVStore once and pass it to `GetPlan`, `SetPlan`, and maybe `SetPlanID` ?

JayT106 commented on x/subscription/keeper/keeper.go:
```diff
@@ -0,0 +1,405 @@
+	}
+
+	k.SetPlan(ctx, plan)
+	k.SetPlanID(ctx, planID+1)
```
Why does the keeper always update planID to the latest planID + 1? What if we just store the ID of the latest plan in `SetPlan` and increase the ID counts only when creating the new plan?

Another question for the KVStore assign cost? Is it heavy to access KVStore by key? If the answer is yes, maybe we can try to reduce the KVStore assign across functions?

yihuang replied:
It's used to generate unique plan id, the value stored in `PlanIdKey` is like the next plan id.

yihuang replied:
> Another question for the KVStore assign cost? Is it heavy to access KVStore by key? If the answer is yes, maybe we can try to reduce the KVStore assign across functions?

Do you mean the cost of `ctx.KVStore(k.storeKey)`, and to share the returned store object?

JayT106 replied:
Yes. After checking your link. It looks good! Thanks.

leejw51crypto commented:
could you squash into one commit?

JayT106 commented on x/subscription/types/key.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+// GetSubscriptionIDFromBytes returns planID in uint64 format from a byte array
+func GetSubscriptionIDFromBytes(bz []byte) (planID uint64) {
```
planID -> subscriptionID ?

yihuang replied:
fixed

JayT106 commented on x/subscription/types/key.go:
```diff
@@ -0,0 +1,124 @@
+	return binary.BigEndian.Uint64(bz)
+}
+
+// GetSubscriptionIDFromBytes returns planID in uint64 format from a byte array
```
planID -> subscriptionID ?

yihuang replied:
fixed

JayT106 commented:
Question for the KVStore in `Keeper`, I guess it uses the KVStore of the cacheMultiStore, so when the keeper sets the data with keys to the KVStore, when these data will be flushed into the disk?

yihuang commented:
> Question for the KVStore in `Keeper`, I guess it uses the KVStore of the cacheMultiStore, so when the keeper sets the data with keys to the KVStore, when these data will be flushed into the disk?

it's [here](https://github.com/cosmos/cosmos-sdk/blob/master/baseapp/abci.go#L293), the abci commit event happens once for each block, at the end of block processing.

yihuang commented:
> could you squash into one commit?

done

leejw51crypto commented:
message MsgCreateSubscriptionResponse {}
how about returning subscription id? like plan_id when creating plan?

tomtau commented on x/subscription/client/cli/tx.go:
```diff
@@ -0,0 +1,274 @@
+			if err != nil {
+				return err
+			}
+			// Get voting address
```
owner address instead of voting address?

yihuang replied:
fixed

tomtau commented on x/subscription/keeper/keeper.go:
```diff
@@ -0,0 +1,405 @@
+
+	var gasPerCollection uint32
+	k.paramSpace.Get(ctx, types.KeyGasPerCollection, &gasPerCollection)
+	ctx.GasMeter().ConsumeGas(sdk.Gas(gasPerCollection), "create subscription")
```
in spec, it says:
```
ConsumeGas( count_period(create_time, expiration_time, plan.cron_spec, plan.cron_tz) * GasPerCollection )
```

but here, it'll just charge a fixed param gasPerCollection?

yihuang replied:
yes, forgot that, added now.

tomtau commented on x/subscription/keeper/keeper.go:
```diff
@@ -0,0 +1,405 @@
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeCreatePlan,
+			sdk.NewAttribute(types.AttributeKeyPlanID, strconv.FormatInt(int64(plan.PlanId), 10)),
```
what happens in these event emitting if PlanId/SubscriptionId doesn't fit in int64?

yihuang replied:
There's actually a `FormatUint`, I just changed to that.
But still, I think when PlanId/SubscriptionId overflow, it'll wrap around, could possibly cause id conflict.
Maybe it's better to stop the service when this happens?

tomtau commented on x/subscription/types/calendar.go:
```diff
@@ -0,0 +1,172 @@
+
+	qCycles := remDays / daysPer4Y
+	if qCycles == 25 {
+		// TODO I think this path is never reached
```
maybe year 2074 or something like that?

yihuang replied:
I tried this one, it don't reach here, I've tried lots of cases. 😄

yihuang commented on proto/subscription/v1/query.proto:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  rpc Params(QueryParamsRequest) returns (QueryParamsResponse) {
+    option (google.api.http).get = "/chainmain/subscription/v1beta1/params";
```
fixed

yihuang commented:
> 1. message MsgCreateSubscriptionResponse {}
>    how about returning subscription id? like plan_id when creating plan?

added

> 2. tzoffset as int32?
>    Plan, MsgCreatePlan,
>    uint32 tzoffset = 7;  // in seconds
>    how about int32? ,
>    because it can be UTC plus or minus ?

good idea, added now.

JayT106 commented:
> > Question for the KVStore in `Keeper`, I guess it uses the KVStore of the cacheMultiStore, so when the keeper sets the data with keys to the KVStore, when these data will be flushed into the disk?
> 
> it's [here](https://github.com/cosmos/cosmos-sdk/blob/master/baseapp/abci.go#L293), the abci commit event happens once for each block, at the end of block processing.

Thanks!

leejw51crypto commented:
abc.go, func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, k keeper.Keeper) { 
..
these times are all utc?
could you comment about it
```
blockTime := uint64(ctx.BlockTime().Unix())
if blockTime >= subscription.NextCollectionTime {		
```

leejw51crypto commented:
repeated cosmos.base.v1beta1.Coin
could you modify to cosmos.base.v1.Coin?

leejw51crypto reviewed (approved):
lgtm

yihuang commented:
> ```
> repeated cosmos.base.v1beta1.Coin
> ```
> 
> could you modify to cosmos.base.v1.Coin?

This namespace is defined in cosmos-sdk, I don't think we can change it ;D

yihuang commented:
> abc.go, func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, k keeper.Keeper) {
> ..
> these times are all utc?
> could you comment about it?
> BlockTime().Unix() unit is seconds?
> 
> ```
> blockTime := uint64(ctx.BlockTime().Unix())
> if blockTime >= subscription.NextCollectionTime {		
> ```

Yeah, these are all unix timestamps in seconds.

JayT106 reviewed (approved):
lgtm

tomtau reviewed (changes requested):
we can probably merge this by Friday -- so may be good to resolve conflicts + changelog entry for unreleased 2.0.0?

tomtau commented:
@yihuang one more test if it can be added -- init a Chain with https://github.com/crypto-org-chain/chain-main/releases/tag/v1.1.0 and then cosmovisor upgrade for v2 with subscriptions? probably some handler to init storage needs to be added?

yihuang commented:
> @yihuang one more test if it can be added -- init a Chain with https://github.com/crypto-org-chain/chain-main/releases/tag/v1.1.0 and then cosmovisor upgrade for v2 with subscriptions? probably some handler to init storage needs to be added?

Done, migration seems happen smoothly.

tomtau commented on CHANGELOG.md:
```diff
@@ -1,6 +1,15 @@
 # Changelog
 
+*March 17, 2021*
```
```suggestion
*unreleased*
```
